src/libasr/asr_utils.h

```diff
@@ -2430,21 +2446,13 @@ inline bool dimension_expr_equal(ASR::expr_t* dim_a, ASR::expr_t* dim_b) {
     if( !(dim_a && dim_b) ) {
         return true;
     }
-    ASR::expr_t* dim_a_fallback = nullptr;
-    ASR::expr_t* dim_b_fallback = nullptr;
-    if( ASR::is_a<ASR::Var_t>(*dim_a) &&
-        ASR::is_a<ASR::Variable_t>(
-            *ASR::down_cast<ASR::Var_t>(dim_a)->m_v) ) {
-        dim_a_fallback = ASRUtils::EXPR2VAR(dim_a)->m_symbolic_value;
-    }
-    if( ASR::is_a<ASR::Var_t>(*dim_b) &&
-        ASR::is_a<ASR::Variable_t>(
-            *ASR::down_cast<ASR::Var_t>(dim_b)->m_v) ) {
-        dim_b_fallback = ASRUtils::EXPR2VAR(dim_b)->m_symbolic_value;
-    }
-    if( !ASRUtils::expr_equal(dim_a, dim_b) &&
-        !(dim_a_fallback && ASRUtils::expr_equal(dim_a_fallback, dim_b)) &&
-        !(dim_b_fallback && ASRUtils::expr_equal(dim_a, dim_b_fallback)) ) {
+    int dim_a_int, dim_b_int;
+    if (ASRUtils::extract_value(ASRUtils::expr_value(dim_a), dim_a_int)
+        && ASRUtils::extract_value(ASRUtils::expr_value(dim_b), dim_b_int)) {
+        return dim_a_int == dim_b_int;
+    }
```

Comment:
I think expr_equal checks for integer constant or real constant. If any other cases are to be handled please add the cases to expr_equal.

Reply:
>I think expr_equal checks for integer constant or real constant

It seems to check for only `IntegerBinOp` and `Var`.

https://github.com/lcompilers/lpython/blob/621ffc09164e52945964bb59f65f491b5a8e024d/src/libasr/asr_utils.h#L2380-L2427

The idea here is to reuse the existing constructs. It might be possible to extend `expr_equal()`, but I think a better approach is to reuse the existing constructs as much as possible.

Reply:
See LFortran

Reply:
Let's look into the possibilities of improving this later. This PR is being lengthier.